File: src/api/player.ts
import { tauriInvoke } from "@api/tauri";

/**
 * 玩家条目
 */
export interface PlayerEntry {
  uuid: string;
  name: string;
}

/**
 * 封禁条目
 */
export interface BanEntry {
  uuid: string;
  name: string;
  reason: string;
  source: string;
  created: string;
  expires: string;
}

/**
 * OP (管理员) 条目
 */
export interface OpEntry {
  uuid: string;
  name: string;
  level: number;
  bypasses_player_limit: boolean;
}

/**
 * 玩家管理 API
 */
export const playerApi = {
  /**
   * 获取白名单 (从文件读取，任何时候都可以)
   */
  async getWhitelist(serverPath: string): Promise<PlayerEntry[]> {
    return tauriInvoke("get_whitelist", { serverPath });
  },

  /**
   * 获取封禁玩家列表
   */
  async getBannedPlayers(serverPath: string): Promise<BanEntry[]> {
    return tauriInvoke("get_banned_players", { serverPath });
  },

  /**
   * 获取 OP 列表
   */
  async getOps(serverPath: string): Promise<OpEntry[]> {
    return tauriInvoke("get_ops", { serverPath });
  },

  /**
   * 添加玩家到白名单 (向运行中的服务器发送命令)
   */
  async addToWhitelist(serverId: string, name: string): Promise<string> {
    return tauriInvoke("add_to_whitelist", { serverId, name });
  },

  /**
   * 从白名单移除玩家
   */
  async removeFromWhitelist(serverId: string, name: string): Promise<string> {
    return tauriInvoke("remove_from_whitelist", { serverId, name });
  },

  /**
   * 封禁玩家
   */
  async banPlayer(serverId: string, name: string, reason: string = ""): Promise<string> {
    return tauriInvoke("ban_player", { serverId, name, reason });
  },

  /**
   * 解封玩家
   */
  async unbanPlayer(serverId: string, name: string): Promise<string> {
    return tauriInvoke("unban_player", { serverId, name });
  },

  /**
   * 添加 OP
   */
  async addOp(serverId: string, name: string): Promise<string> {
    return tauriInvoke("add_op", { serverId, name });
  },

  /**
   * 移除 OP
   */
  async removeOp(serverId: string, name: string): Promise<string> {
    return tauriInvoke("remove_op", { serverId, name });
  },

  /**
   * 踢出玩家
   */
  async kickPlayer(serverId: string, name: string, reason: string = ""): Promise<string> {
    return tauriInvoke("kick_player", { serverId, name, reason });
  },

  /**
   * 导出日志
   */
  async exportLogs(logs: string[], savePath: string): Promise<void> {
    return tauriInvoke("export_logs", { logs, savePath });
  },
};

File: backend/tauri-host/src/adapters/http/command_registry/player.rs
use super::common::{CommandHandler, RegistryBuilder, parse_params};
use super::requests::{
    BanPlayerRequest, ExportLogsRequest, KickPlayerRequest, PlayerActionRequest, ServerPathRequest,
};
use crate::commands::server::players as player_commands;
use serde_json::Value;
pub(super) fn register_handlers(builder: &mut RegistryBuilder) {
    builder.register("get_whitelist", handle_get_whitelist as CommandHandler);
    builder.register("get_banned_players", handle_get_banned_players as CommandHandler);
    builder.register("get_ops", handle_get_ops as CommandHandler);
    builder.register("add_to_whitelist", handle_add_to_whitelist as CommandHandler);
    builder.register(
        "remove_from_whitelist",
        handle_remove_from_whitelist as CommandHandler,
    );
    builder.register("ban_player", handle_ban_player as CommandHandler);
    builder.register("unban_player", handle_unban_player as CommandHandler);
    builder.register("add_op", handle_add_op as CommandHandler);
    builder.register("remove_op", handle_remove_op as CommandHandler);
    builder.register("kick_player", handle_kick_player as CommandHandler);
    builder.register("export_logs", handle_export_logs as CommandHandler);
}

fn handle_get_whitelist(
    params: Value,
) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: ServerPathRequest = parse_params(params)?;
        let result = player_commands::get_whitelist(req.server_path)?;
        serde_json::to_value(result).map_err(|e| e.to_string())
    })
}

fn handle_get_banned_players(
    params: Value,
) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: ServerPathRequest = parse_params(params)?;
        let result = player_commands::get_banned_players(req.server_path)?;
        serde_json::to_value(result).map_err(|e| e.to_string())
    })
}

fn handle_get_ops(params: Value) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: ServerPathRequest = parse_params(params)?;
        let result = player_commands::get_ops(req.server_path)?;
        serde_json::to_value(result).map_err(|e| e.to_string())
    })
}

fn handle_add_to_whitelist(
    params: Value,
) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: PlayerActionRequest = parse_params(params)?;
        let result = player_commands::add_to_whitelist(req.server_id, req.name)?;
        Ok(Value::String(result))
    })
}

fn handle_remove_from_whitelist(
    params: Value,
) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: PlayerActionRequest = parse_params(params)?;
        let result = player_commands::remove_from_whitelist(req.server_id, req.name)?;
        Ok(Value::String(result))
    })
}

fn handle_ban_player(params: Value) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: BanPlayerRequest = parse_params(params)?;
        let result = player_commands::ban_player(req.server_id, req.name, req.reason)?;
        Ok(Value::String(result))
    })
}

fn handle_unban_player(
    params: Value,
) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: PlayerActionRequest = parse_params(params)?;
        let result = player_commands::unban_player(req.server_id, req.name)?;
        Ok(Value::String(result))
    })
}

fn handle_add_op(params: Value) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: PlayerActionRequest = parse_params(params)?;
        let result = player_commands::add_op(req.server_id, req.name)?;
        Ok(Value::String(result))
    })
}

fn handle_remove_op(params: Value) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: PlayerActionRequest = parse_params(params)?;
        let result = player_commands::remove_op(req.server_id, req.name)?;
        Ok(Value::String(result))
    })
}

fn handle_kick_player(params: Value) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: KickPlayerRequest = parse_params(params)?;
        let result = player_commands::kick_player(req.server_id, req.name, req.reason)?;
        Ok(Value::String(result))
    })
}

fn handle_export_logs(params: Value) -> futures::future::BoxFuture<'static, Result<Value, String>> {
    Box::pin(async move {
        let req: ExportLogsRequest = parse_params(params)?;
        player_commands::export_logs(req.logs, req.save_path)?;
        Ok(Value::Null)
    })
}
